core/shared/src/main/scala/fs2/text.scala

```diff
@@ -306,11 +326,111 @@ object text {
 
   /** Encodes a stream of `String` in to a stream of bytes using the given charset. */
   def encode[F[_]](charset: Charset): Pipe[F, String, Byte] =
-    _.mapChunks(c => c.flatMap(s => Chunk.array(s.getBytes(charset))))
+    encodeC(charset)(_).unchunks
 
   /** Encodes a stream of `String` in to a stream of `Chunk[Byte]` using the given charset. */
-  def encodeC[F[_]](charset: Charset): Pipe[F, String, Chunk[Byte]] =
-    _.mapChunks(_.map(s => Chunk.array(s.getBytes(charset))))
+  def encodeC[F[_]](charset: Charset): Pipe[F, String, Chunk[Byte]] = {
+    def encodeC(
+        encoder: CharsetEncoder,
+        acc: Chunk[Char],
+        s: Stream[F, Chunk[Char]]
+    ): Pull[F, Chunk[Byte], Unit] =
+      s.pull.uncons1.flatMap { r =>
+        val toEncode = r match {
+          case Some((c, _)) => acc ++ c
+          case None         => acc
+        }
+
+        val isLast = r.isEmpty
+        val outBufferSize =
+          math.max(encoder.maxBytesPerChar(), encoder.averageBytesPerChar() * toEncode.size).toInt
+
+        val out = ByteBuffer.allocate(outBufferSize)
+
+        val inBuffer = toEncode.toCharBuffer
+        encoder.encode(inBuffer, out, isLast)
+        (out: Buffer).flip()
+
+        val nextAcc =
+          if (inBuffer.remaining() > 0) Chunk.charBuffer(inBuffer.slice()) else Chunk.empty
+
+        val rest = r match {
+          case Some((_, tail)) => tail
+          case None            => Stream.empty
+        }
+
+        if (out.remaining() > 0) {
+          Pull.output1(Chunk.ByteBuffer.view(out)) >> encodeC(encoder, nextAcc, rest)
+        } else if (!isLast) {
+          encodeC(encoder, nextAcc, rest)
+        } else if (nextAcc.nonEmpty) {
+          encodeC(encoder, nextAcc, rest)
+        } else flush(encoder, ByteBuffer.allocate(0))
+      }
+
+    @tailrec
+    def flush(
+        encoder: CharsetEncoder,
+        out: ByteBuffer
+    ): Pull[F, Chunk[Byte], Unit] = {
+      (out: Buffer).clear()
+      encoder.flush(out) match {
+        case res if res.isUnderflow =>
+          if (out.position() > 0) {
+            (out: Buffer).flip()
+            Pull.output1(Chunk.ByteBuffer.view(out)) >> Pull.done
+          } else
+            Pull.done
+        case res if res.isOverflow =>
+          val newSize = (out.capacity + encoder.maxBytesPerChar() * 2).toInt
+          val bigger = ByteBuffer.allocate(newSize)
+          flush(encoder, bigger)
+        case res =>
+          ApplicativeThrow[Pull[F, Chunk[Byte], *]].catchNonFatal(res.throwException())
+      }
+    }
+
+    def cutBOMs(
+        s: Stream[F, String],
+        bom: Seq[Byte],
+        doDrop: Boolean
+    ): Pull[F, Chunk[Byte], Unit] =
+      s.pull.uncons1.flatMap {
+        case Some((hd, tail)) =>
+          val bytes = Chunk.array(hd.getBytes(charset))
+          val dropped = if (doDrop && bytes.startsWith(bom)) bytes.drop(bom.length) else bytes
```

Comment:
`.startsWith` is not particularly optimized, esp. since we are comparing seqs with boxed bytes. No idea if it adds up to something meaningful.